src/sentry/db/models/fields/node.py

```diff
@@ -158,45 +191,53 @@ def on_delete(self, instance, **kwargs):
         nodestore.delete(value.id)
 
     def to_python(self, value):
-        if isinstance(value, six.string_types) and value:
+        # Case 1: We were passed something empty or None, with no idea what ID
+        # to store it under. Set it to an empty dict and generate a default.
+        if not value:
+            value = {}
+            node_id = self.id_func(value)
+
+        # Case 2: This is a value we've loaded from the database, it should be
+        # decompressed/unpickled, and we should end up with {"node_id": "XYZ"}.
+        # Extract the node_id.
+        elif isinstance(value, six.string_types):
             try:
                 value = pickle.loads(decompress(value))
+                node_id = value.pop('node_id')
+                assert value == {}  # shouldn't be anything else left in there.
+                value = None  # This tells NodeData we need to load the data.
             except Exception as e:
+                # TODO this is a bit dangerous as a failure to read/decode the
+                # node_id will end up with this record being replaced with an
+                # empty value under a new key, potentially orphaning an
+                # original value in nodestore. OTOH if we can't decode the info
+                # here, the node was already effectively orphaned.
```

Comment:
Should we cleanup the orphaned record?

Reply:
If this data is actually borked and we can't decode the `node_id` we have no idea where to find the orphaned record, or even if there is one.

I guess the main danger that I wanted to warn against was deleting/orphaning the object in the case of some transient error when the data was actually fine. I _think_ pickle.loads and decompress should only fail in cases where the data is actually bad, and not because of, say, a ConnectionTimeout`, but not 100% sure.

Reply:
For what it's worth, we have a global TTL on Riak nodestore in production, and for other that run Django nodestore, there is a query in cleanup that doesn't reference node objects by their IDs.